python/lsst/ts/ess/labjack/base_labjack_data_client.py

```diff
@@ -0,0 +1,177 @@
+    """
+
+    def __init__(
+        self,
+        config: types.SimpleNamespace,
+        topics: salobj.Controller | types.SimpleNamespace,
+        log: logging.Logger,
+        simulation_mode: int = 0,
+    ) -> None:
+        # handle to LabJack device
+        self.handle = None
+
+        # The thread pool executor used by `run_in_thread`.
+        self._thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+
+        super().__init__(
+            config=config, topics=topics, log=log, simulation_mode=simulation_mode
+        )
+
+    @property
+    def connected(self) -> bool:
+        return self.handle is not None
+
+    def descr(self) -> str:
+        return f"identifier={self.config.identifier}"
+
+    async def run_in_thread(self, func: Callable[[], Any], timeout: float) -> Any:
+        """Run a blocking function in a thread pool executor.
+
+        Only one function will run at a time, because all calls use the same
+        thread pool executor, which only has a single thread.
+
+        Parameters
+        ----------
+        func : `Callable`
+            The blocking function to run;
+            The function must take no arguments.
+            For example: ``self._blocking_connect``.
+        timeout : `float`
+            Time limit (seconds).
+        """
+        loop = asyncio.get_running_loop()
+        try:
+            return await asyncio.wait_for(
+                loop.run_in_executor(self._thread_pool, func), timeout=timeout
+            )
+        except Exception:
+            self.log.exception(f"Blocking function {func} failed")
+            raise
+
+    async def connect(self) -> None:
+        """Connect to the LabJack.
+
+        Disconnect first, if connected.
+
+        Raises
+        ------
+        RuntimeError
+            If we cannot read the data.
+        """
+        self.log.info(
+            f"Connect to LabJack {self.config.device_type}, "
+            f"config.identifier={self.config.identifier!r}, "
+            f"config.connection_type={self.config.connection_type}"
+        )
+        await self.run_in_thread(func=self._blocking_connect, timeout=CONNECT_TIMEOUT)
+
+    async def disconnect(self) -> None:
+        """Disconnect from the LabJack. A no-op if disconnected."""
+        try:
+            await self.run_in_thread(
+                func=self._blocking_disconnect, timeout=CONNECT_TIMEOUT
+            )
+        finally:
+            self.handle = None
+
+    def _blocking_connect(self) -> None:
+        """Connect to the LabJack.
+
+        Disconnect first, if connected.
+
+        Call in a thread to avoid blocking the event loop.
+        """
+        if self.handle is not None:
+            self.log.warning("Already connected; disconnecting and reconnecting")
+            self._blocking_disconnect()
+
+        if self.simulation_mode == 0:
+            identifier = self.config.identifier
+            if self.config.connection_type in {"TCP", "WIFI"}:
```

Comment:
What happens in case the connection type is not one of the stated ones? This will silently not connect if left as is.

Reply:
This if statement expands TCP/IP addresses; if the identifier is not a TCP/IP address then it is left unchanged from the configuration.